### ISSUE TITLE
trainmodel.py

### ISSUE THREAD

**Issue by DNikhilesh122** [DNikhilesh122/seq-to-seq-chatbot#3]
import torch
import torch.nn as nn
import pandas as pd
import random
from preprocess import *
from model import Encoder, Decoder

df = pd.read_csv("data/faq.csv")

pairs = [(clean_text(q), clean_text(a)) for q, a in zip(df["question"], df["answer"])]

vocab = Vocabulary()
for q, a in pairs:
    vocab.add_sentence(q)
    vocab.add_sentence(a)

encoder = Encoder(vocab.n_words)
decoder = Decoder(vocab.n_words)

optimizer = torch.optim.Adam(list(encoder.parameters()) + list(decoder.parameters()))
loss_fn = nn.CrossEntropyLoss()

for epoch in range(20):
    total_loss = 0
    for q, a in pairs:

        input_tensor = sentence_to_tensor(vocab, q)
        target_tensor = sentence_to_tensor(vocab, a)

        optimizer.zero_grad()

        encoder_outputs, hidden, cell = encoder(input_tensor.unsqueeze(0))

        decoder_input = torch.tensor([vocab.word2idx[SOS]])

        loss = 0

        for t in range(len(target_tensor)):
            output, hidden, cell = decoder(decoder_input, hidden, cell, encoder_outputs)

            loss += loss_fn(output, target_tensor[t].unsqueeze(0))

            teacher = random.random() < 0.5
            top1 = output.argmax(1)
            decoder_input = target_tensor[t] if teacher else top1

        loss.backward()
        optimizer.step()

        total_loss += loss.item()

    print("Epoch:", epoch, "Loss:", total_loss)

torch.save(encoder.state_dict(), "models/encoder.pth")
torch.save(decoder.state_dict(), "models/decoder.pth")
